### PR TITLE
Warn about more ignored bounds in type aliases

### PR DESCRIPTION
**Pull request by RalfJung**
It seems that all bounds in type aliases are entirely ignored, not just type bounds. This extends the warning appropriately.

I assume this should be made a hard error with the next epoch? I can't see any reason to accept these programs. (And suddenly enforcing these type bounds would be a breaking change.)

### CODE REVIEW

File: src/test/ui/param-bounds-ignored.stderr [hunk 1]
```diff
@@ -0,0 +1,18 @@
+warning[E0122]: bounds are ignored in type aliases
```
Comment (petrochenkov) [inline src/test/ui/param-bounds-ignored.stderr]: Nit: maybe "generic bounds" instead just "bounds"?
Just "bounds" looks like something is missing to me. Not sure.

**Comment by petrochenkov** [comment]
r=me after maybe fixing the nit
@bors delegate+

**Comment by RalfJung** [comment]
Oh fancy, I had no idea bors supports delegation like this. oO

I didn't want to write "type and lifetime bounds", hence my choice of words. I like "generic bounds", thanks for the suggestion!

**Comment by RalfJung** [comment]
@bors r=@petrochenkov 

Let's see if this works...

**Comment by RalfJung** [comment]
Oh, seems like you don't usually put at `@` after `r=`... is that a problem?

**Comment by petrochenkov** [comment]
@RalfJung 
IIUC, the right side of `@bors r=` is for humans only and can contain anything if the comment author has r+ rights.

**Comment by RalfJung** [comment]
*lol* indeed. Now poor bors is confused.

@bors r=petrochenkov

**Comment by bluss** [comment]
Just to be on the safe side, if you propose making it an error, what will happen to code like this, will it continue to be accepted?:

```rust
use std::ops::Index;
type Output<I, T> = <I as Index<T>>::Output;
```

**Comment by RalfJung** [comment]
I think so -- this checks bounds in the binders of type aliases, not touching the code it expands to. I assume you bring this up because it actually works? All I'm advocating for is the compiler tell us when we write something that it doesn't support.

**Comment by bluss** [comment]
Yes it works, but my intuition would be that the bound is required in the type alias definition there, and the compiler is doing the opposite, enforcing the form without bound.

**Comment by RalfJung** [comment]
The current behavior of type aliases is essentially a syntactic expansion. It's more like a macro. Type bounds are checked later. (That's my interpretation anyway, please correct me if that's wrong.)

I do agree that this is somewhat unexpected. However, changing it is backwards-incompatible. So I figured we could at least try to reduce the confusion by telling people when we ignore what they write, e.g. if they do
```
type SendableVec<T: Send> = Vec<T>
```
which actually doesn't mean the `Vec` is sendable.

I suppose what you're saying is that instead of forcing (well, linting) the bounds to be empty, we could lint them to exactly match whatever is required to make sense of the type in the right? I agree that'd be somewhat nicer and more compatible with an eventual future where these bounds are actually checked (at def'n time) and enforced. Unfortunately, I have no idea how to go about implementing that; it seems pretty complicated. Do you think this is a realistic thing to do, and we should rather abort/revert this MR to not have people that want to adhere to the lint change their code twice?

Notice however that for type bounds, we already warn. This just adds the same warning to lifetime bounds.

**Comment by bluss** [comment]
In the long run I think the bound requirement is correct, but this has already turned quite the other way (with regular bounds), so it doesn't matter to stop just this PR.

**Pull request merged**